src/engine/Sort.cpp

```diff
@@ -54,121 +58,148 @@ std::string Sort::getDescriptor() const {
 
-  // Create a unique temporary filename for the external sorter in the index
-  // directory.
+  // Create a unique temporary filename in the index directory.
   const std::string& onDiskBase =
       getExecutionContext()->getIndex().getOnDiskBase();
   ad_utility::UuidGenerator uuidGen;
   std::string tempFilename =
       absl::StrCat(onDiskBase, ".sort.", uuidGen(), ".dat");
 
-  // Get memory limit from the same runtime parameters that we use for the
-  // materialized view writer.
-  //
-  // TODO: Give this runime parameter a more general name.
   ad_utility::MemorySize memoryLimit =
-      getRuntimeParameter<&RuntimeParameters::materializedViewWriterMemory_>();
-
-  // Get the number of columns from the subtree (this works for both lazy and
-  // materialized results).
+      getRuntimeParameter<&RuntimeParameters::sortInMemoryThreshold_>();
   size_t numColumns = subtree_->getResultWidth();
 
-  // Create the external sorter with our dynamic column comparator. We use a
-  // unique_ptr because the Sorter is not moveable (contains std::atomic), but
-  // unique_ptr itself can be moved into the lambda.
-  using Sorter = ad_utility::CompressedExternalIdTableSorter<SortByColumns, 0>;
+  // We use a `unique_ptr` because the `Sorter` is not moveable (contains
+  // `std::atomic`), but the `unique_ptr` can be moved into the lambda below.
   auto sorter = std::make_unique<Sorter>(
       tempFilename, numColumns, memoryLimit, allocator(),
       ad_utility::DEFAULT_BLOCKSIZE_EXTERNAL_ID_TABLE,
       SortByColumns{sortColumnIndices_});
 
-  // Push all rows into the sorter, handling both lazy and materialized results.
-  // For materialized results, we copy the local vocab, for lazy results we
-  // merge the local vocabs from the individual blocks.
-  LocalVocab mergedLocalVocab;
-  if (subRes->isFullyMaterialized()) {
-    sorter->pushBlock(subRes->idTable());
-    mergedLocalVocab = subRes->getCopyOfLocalVocab();
-  } else {
-    for (Result::IdTableVocabPair& pair : subRes->idTables()) {
-      checkCancellation();
-      sorter->pushBlock(std::move(pair.idTable_));
-      mergedLocalVocab.mergeWith(pair.localVocab_);
-    }
+  // Push already collected blocks.
+  for (auto& block : collectedBlocks) {
+    sorter->pushBlock(std::move(block));
+  }
+
+  // Process remaining blocks.
+  while (it != end) {
+    checkCancellation();
+    auto& idTableAndLocalVocab = *it;
+    sorter->pushBlock(std::move(idTableAndLocalVocab.idTable_));
+    mergedLocalVocab.mergeWith(idTableAndLocalVocab.localVocab_);
+    ++it;
   }
 
-  // If laziness is not requested, materialize the result. The `sorter` the
-  // size of the result, so we can reserve exactly the right amount of space
-  // in advance.
+  // If laziness is not requested, materialize the result. The sorter knows the
+  // size of the result, so we can reserve exactly the right amount of space.
   if (!requestLaziness) {
     IdTable result{numColumns, allocator()};
```

Comment:
For this case to be memory efficient (and provide an actual benefit over the in memory sorting), you need to discard the current result. Otherwise you force QLever to keep the original result around (because the reference counter can never reach zero). You can achieve this by passing the result as a value to this function and reset it (or assign the nullptr) before this if clause.